### PR TITLE
WIP - Unify boolean expression and value expression grammars

### PR DESCRIPTION
**Pull request by bbeaudreault**
@sougou I still need to add some tests to this, but wanted to push up what I had to run the existing full test suite as it stands. The 3 main parser test packages pass locally.

When I implemented your example for handling ParenExpr, it gave me some type mismatches throughout the go-code.  So I cleaned that up a bit as well (at least I hope I improved it). Basically, there now is two tuple-related interfaces:

ColTuple (as existed)
ValTuple (converted to an interface)

The old ValTuple was replaced with RowTuple, and made to implement ValTuple. ParenExpr also impelements ValTuple. You'll see in the diff :)

Also I think I need to remove your change to the Makefile -- I don't have goyacc.

### CODE REVIEW

File: go/vt/sqlparser/ast.go [hunk 1]
```diff
@@ -1276,51 +1264,38 @@ func (node *ColName) Equal(c *ColName) bool {
...
 
+// ValTuple represents a list of value expressions.
+// It can be RowTuple or ParenExpr.
+type ValTuple interface {
+	iRowTuple()
+	Expr
+}
+
+func (RowTuple) iRowTuple()  {}
+func (ParenExpr) iRowTuple() {}
```
Comment (sougou) [inline go/vt/sqlparser/ast.go]: I actually don't think this is needed. I think everything will compile fine if you just deleted this interface.

Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: Just tried it and deleting it doesn't compile. Definitely open to suggestions. I needed this to support the new if statement we added in sql.y -- the problem is:

row_tuple returned either a ParenExpr or a ValTuple, depending on the size of the value_expression_list. So I need a supertype that both of those fit into, to define as the return type of row_tuple.

Comment (sougou) [inline go/vt/sqlparser/ast.go]: I think it should work if you changed the return type of row_tuple to expr. I'll try to play around with it if that doesn't work.

Comment (sougou) [inline go/vt/sqlparser/ast.go]: Take a look at this diff: https://github.com/sougou/vitess/commit/11c55ddddce8c89b973ee8a3e3f1a0c4d5ac45f3. This one compiles for me.

Comment (bbeaudreault) [inline go/vt/sqlparser/ast.go]: Yep, you're right. Expr already is a supertype of both ParenExpr and RowTupl. Integrated your patch, and removed the now unnecessary diffs of renaming ValTuple to RowTuple.

Gonna add some tests now

**Comment by bbeaudreault** [comment]
Tests are failing due to a failure to download mariadb. They were passing before.  I'll be pushing more tests in the next couple of hours, which should result in a successful re-run.

**Comment by bbeaudreault** [comment]
@sougou I added a bunch of tests in parser, plan builder, and endtoend. Let me know if you'd like more in any specific area/cases. Otherwise I think this is ready for a final review?

**Comment by bbeaudreault** [comment]
Though it looks like it's still failing to download mariadb...

**Comment by bbeaudreault** [comment]
The tests are broken for a grpc issue. I cant run the full suite locally, so I guess we can wait for that fix before moving forward unless you're feeling generous :)

**Comment by michael-berlin** [comment]
We had periods of broken Travis runs before and used an honor system before.

Therefore, let's stay pragmatic here: If you tell us that the tests are passing locally, we'll believe you and merge this. No need to halt development.

**Comment by bbeaudreault** [comment]
Thanks @michael-berlin, I appreciate the pragmatism :).

@sougou looks like tests are working again and passing. I think this is ready for a final review, when you're ready.

**Comment by sougou** [comment]
LGTM
Nice work. I can merge after you resolve the conflicts.

**Comment by bbeaudreault** [comment]
Resolved all conflicts and all tests are passing. Thanks @sougou!

**Comment by bbeaudreault** [comment]
:man_dancing:

**Pull request merged**